src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.prev = prev;
+            this.item = element;
+            this.next = next;
+        }
+    }
+
+    private void checkIndex(int index, int size) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index is not valid. Index: "
+                    + index + ". Size of list: " + size);
+        }
+    }
+
+    private void addMiddle(T value, int index) {
+        Node middleNode = new Node(findNode(index - 1), value, findNode(index));
+        findNode(index - 1).next = middleNode;
+        findNode(index).prev = middleNode;
+        size++;
+    }
+
+    private void addLast(T value) {
+        Node lastNode = new Node(last, value, null);
+        last = lastNode;
+        if (lastNode.prev != null) {
+            lastNode.prev.next = lastNode;
+        }
+        size++;
+        if (first == null) {
+            first = lastNode;
+        }
+    }
+
+    private void addFirst(T value) {
+        Node firstNode = new Node(null, value, first);
+        first = firstNode;
+        if (firstNode.next != null) {
+            firstNode.next.prev = firstNode;
+        }
+        size++;
+        if (last == null) {
+            last = firstNode;
+        }
+    }
+
+    private Node findNode(int index) {
+        if (index < size / 2) {
+            Node getNode = first;
+            for (int i = 0; i <= size / 2; i++) {
+                if (i == index) {
+                    return getNode;
+                }
+                getNode = getNode.next;
+            }
+        }
+        Node getNode = last;
+        for (int i = size - 1; i >= size / 2; i--) {
+            if (i == index) {
+                return getNode;
+            }
+            getNode = getNode.prev;
+        }
+        return null;
+    }
+
+    private void unlink(Node node, int index) {
```

Comment:
```suggestion
    private void unlink(Node<T> node) {
```
you can compare references 
if node == head it means that index = 0
if node == tail it means that it is the last node